vyper/context/types/union.py

```diff
@@ -0,0 +1,70 @@
+from vyper.exceptions import (
+    InvalidOperation,
+    VyperException,
+)
+
+
+class UnionType(set):
+    """
+    Set subclass for literal values where the final type is not yet determined.
+
+    When this object is compared to another type, invalid types for the comparison
+    are removed. For example, the literal `1` is initially a `UnionType` of
+    `{int128, uint256}`. If the type is then compared to `-1` it is considered to
+    be `int128` and subsequent comparisons to `uint256` will return `False`.
```

Comment:
This is some crazy statefulness. If I compare something to it, it changes? This will be difficult to debug.